### ISSUE TITLE
iso image with systemd stage1 fails to boot due to nonexistent `/dev/root`

### ISSUE THREAD

**Issue by anna328p** [NixOS/nixpkgs#327187]
### Describe the bug

An iso image built with systemd stage1 (`initrd.systemd.enable = true`) fails to boot due to nonexistent `/dev/root`.

`/etc/fstab`:
```
overlay /nix/store overlay x-initrd.mount,lowerdir=/nix/.ro-store,upperdir=/nix/.rw-store/store,workdir=/nix/.rw-store/work 0 0
```

```
bash-5.2# systemctl status sysroot-iso.mount |cat
× sysroot-iso.mount - /sysroot/iso
     Loaded: loaded (/nix/store/4kazmgs52j4zp0kr3h5y9vdpmv6kc00d-initrd-fstab; generated)
     Active: failed (Result: exit-code) since Sun 2024-07-14 19:16:40 UTC; 1min 7s ago
      Where: /sysroot/iso
       What: /dev/root
       Docs: man:fstab(5)
             man:systemd-fstab-generator(8)
        CPU: 1ms

Jul 14 19:16:40 localhost systemd[1]: Mounting /sysroot/iso...
Jul 14 19:16:40 localhost systemd[1]: sysroot-iso.mount: Mount process exited, code=exited, status=1/FAILURE
Jul 14 19:16:40 localhost mount[172]: mount: /sysroot/iso: no valid filesystem type specified.
Jul 14 19:16:40 localhost systemd[1]: sysroot-iso.mount: Failed with result 'exit-code'.
Jul 14 19:16:40 localhost systemd[1]: Failed to mount /sysroot/iso.
```

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

 - system: `"x86_64-linux"`
 - host os: `Linux 6.7.10, NixOS, 24.05 (Uakari), 24.05.20240317.f471be9`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.23.1`
 - nixpkgs: `/nix/store/xb5p9kg3gy00q7c8s638jf15j6qqj473-source`

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by nvmd** [comment]
Experiencing the same issue but via nix-community/nixos-images#258.

full contents of `fstab`:

```
tmpfs / tmpfs x-initrd.mount,mount=0755 0 0
/dev/root /iso auto x-initrd.mount 0 0
/iso/nix-store.squashfs /nix/.ro-store squashfs x-initrd.mount,loop 0 0
tmpfs /nix/.rw-store tmpfs x-initrd.mount,mode=0755 0 0
overlay /nix/store overlay x-initrd.mount,lowerdir=/nix/.ro-store,upperdir=/nix/.rw-store/store,workdir=/nix/.rw-store/work 0 0
```

**Comment by anna328p** [comment]
I have no idea why I'd only ended up posting the last line, I swear I copied the whole thing lol

**Comment by Mic92** [comment]
I think this broke the netboot image: https://github.com/NixOS/nixpkgs/pull/331712

**Comment by ElvishJerricco** [comment]
#291750, which is blocked by #309190

**Comment by nvmd** [comment]
Also, almost a duplicate of this issue, with some background information on `root=` kernel command line option handling #217173

**Comment by ElvishJerricco** [comment]
Closed by #291750
